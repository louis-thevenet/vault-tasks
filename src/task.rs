use anyhow::bail;
use chrono::{NaiveDate, NaiveDateTime};
use core::fmt;
use log::{debug, info};
use std::{
    fmt::Display,
    fs::{read_to_string, File},
    io::Write,
};

use crate::config::Config;

#[derive(Debug, Hash, Eq, PartialEq, Clone)]
pub enum State {
    Done,
    ToDo,
}
impl Display for State {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Done => write!(f, "Done")?,
            Self::ToDo => write!(f, "To Do")?,
        }
        Ok(())
    }
}
#[derive(Debug, Hash, Eq, PartialEq, Clone)]
/// This type accounts for the case where the task has a due date but no exact due time
pub enum DueDate {
    NoDate,
    Day(NaiveDate),
    DayTime(NaiveDateTime),
}
impl Display for DueDate {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Day(date) => write!(f, "{date}"),
            Self::DayTime(date) => write!(f, "{date}"),
            Self::NoDate => Ok(()),
        }
    }
}

impl DueDate {
    pub fn to_string_format(&self, config: &Config) -> String {
        let format_date = if config.use_american_format.unwrap_or(true) {
            "%Y/%m/%d"
        } else {
            "%d/%m/%Y"
        };
        let format_datetime = if config.use_american_format.unwrap_or(true) {
            "%Y/%m/%d %T"
        } else {
            "%d/%m/%Y %T"
        };

        match self {
            Self::Day(date) => date.format(format_date).to_string(),
            Self::DayTime(date) => date.format(format_datetime).to_string(),
            Self::NoDate => String::new(),
        }
    }
}

#[derive(Debug, Hash, Eq, PartialEq, Clone)]
pub struct Task {
    pub due_date: DueDate,
    pub name: String,
    pub priority: usize,
    pub state: State,
    pub description: Option<String>,
    pub tags: Option<Vec<String>>,
    pub line_number: usize,
}

impl Default for Task {
    fn default() -> Self {
        Self {
            due_date: DueDate::NoDate,
            name: String::from("New Task"),
            priority: 0,
            state: State::ToDo,
            tags: None,
            description: None,
            line_number: 1,
        }
    }
}

impl fmt::Display for Task {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        writeln!(f, "Name: \t -> {0}", self.name)?;

        match self.due_date {
            DueDate::NoDate => (),
            DueDate::DayTime(date) => writeln!(f, "Due: \t -> {date}")?,
            DueDate::Day(date) => writeln!(f, "Due: \t {date}")?,
        }

        writeln!(f, "State: \t -> {}", self.state)?;

        if self.priority > 0 {
            writeln!(f, "Prio.: \t -> {}", self.priority)?;
        }

        if let Some(description) = &self.description {
            writeln!(f, "Desc.:\t -> \"{description}\"")?;
        }
        if let Some(tags) = &self.tags {
            writeln!(f, "Tags: \t -> {tags:?}")?;
        }
        fmt::Result::Ok(())
    }
}
impl Task {
    fn get_fixed_attributes(&self, config: &Config, indent_length: usize) -> String {
        let indent = " ".repeat(indent_length);

        let state_str = match self.state {
            State::Done => "X",
            State::ToDo => " ",
        };

        let priority = if self.priority > 0 {
            format!("p{} ", self.priority)
        } else {
            String::new()
        };

        let mut due_date = self.due_date.to_string_format(config);
        if !due_date.is_empty() {
            due_date.push(' ');
        }

        let tags_str = self.tags.as_ref().map_or_else(String::new, |tags| {
            tags.clone()
                .iter()
                .map(|t| format!("#{t}"))
                .collect::<Vec<String>>()
                .join(" ")
        });

        format!(
            "{}- [{}] {} {}{}{}",
            indent, state_str, self.name, due_date, priority, tags_str
        )
    }

    pub fn fix_task_attributes(&self, config: &Config, filename: &str) -> anyhow::Result<()> {
        let mut path = config.vault_path.clone();
        path.push(filename);
        let content = read_to_string(path.clone())?;
        let mut lines = content.split('\n').collect::<Vec<&str>>();

        if lines.len() < self.line_number - 1 {
            bail!(
                "Task's line number {} was greater than length of file {}",
                self.line_number,
                filename
            );
        }

        let indent_length = lines[self.line_number - 1]
            .chars()
            .take_while(|c| c.is_whitespace())
            .count();

        let fixed_line = self.get_fixed_attributes(config, indent_length);

        if lines[self.line_number - 1] != fixed_line {
            debug!(
                "\nReplacing\n{}\nWith\n{}\n",
                lines[self.line_number - 1],
                self.get_fixed_attributes(config, indent_length,)
            );
            lines[self.line_number - 1] = &fixed_line;

            let mut file = File::create(path)?;
            file.write_all(lines.join("\n").as_bytes())?;

            info!("Wrote to {filename} at line {}", self.line_number);
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_fix_attributes() {
        let config = Config::default();
        let task = Task {
            due_date: DueDate::Day(NaiveDate::from_ymd_opt(2021, 12, 3).unwrap()),
            name: String::from("Test Task"),
            priority: 1,
            state: State::ToDo,
            tags: Some(vec![String::from("tag1"), String::from("tag2")]),
            description: Some(String::from("This is a test task.")),
            line_number: 2,
        };
        let res = task.get_fixed_attributes(&config, 0);
        assert_eq!(res, "- [ ] Test Task 2021/12/03 p1 #tag1 #tag2");
    }

    #[test]
    fn test_fix_attributes_with_no_date() {
        let config = Config::default();
        let task = Task {
            due_date: DueDate::NoDate,
            name: String::from("Test Task with No Date"),
            priority: 2,
            state: State::Done,
            tags: Some(vec![String::from("tag3")]),
            description: None,
            line_number: 3,
        };

        let res = task.get_fixed_attributes(&config, 0);
        assert_eq!(res, "- [X] Test Task with No Date p2 #tag3");
    }
}
