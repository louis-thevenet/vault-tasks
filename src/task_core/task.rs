use chrono::{NaiveDate, NaiveDateTime};
use color_eyre::{eyre::bail, Result};
use core::fmt;
use std::{
    fmt::Display,
    fs::{read_to_string, File},
    io::Write,
    path::PathBuf,
};
use tracing::{debug, info};

use crate::config::Config;

const STATE_TO_DO_EMOJI: &str = "‚ùå";
const STATE_DONE_EMOJI: &str = "‚úÖ";
pub const DUE_DATE_EMOJI: &str = "üìÖ";
pub const PRIORITY_EMOJI: &str = "‚ùó";
#[derive(Debug, Hash, Eq, PartialEq, Clone)]
pub enum State {
    Done,
    ToDo,
}
impl Display for State {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Done => write!(f, "{STATE_DONE_EMOJI}")?,
            Self::ToDo => write!(f, "{STATE_TO_DO_EMOJI}")?,
        }
        Ok(())
    }
}
#[derive(Debug, Hash, Eq, PartialEq, Clone)]
/// This type accounts for the case where the task has a due date but no exact due time
pub enum DueDate {
    NoDate,
    Day(NaiveDate),
    DayTime(NaiveDateTime),
}
impl Display for DueDate {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Day(date) => write!(f, "{date}"),
            Self::DayTime(date) => write!(f, "{date}"),
            Self::NoDate => Ok(()),
        }
    }
}

impl DueDate {
    pub fn to_display_format(&self, not_american_format: bool) -> String {
        if matches!(self, Self::NoDate) {
            String::new()
        } else {
            format!(
                "{DUE_DATE_EMOJI} {}",
                self.to_string_format(not_american_format)
            )
        }
    }
    pub fn to_string_format(&self, not_american_format: bool) -> String {
        let format_date = if !not_american_format {
            "%Y/%m/%d"
        } else {
            "%d/%m/%Y"
        };
        let format_datetime = if !not_american_format {
            "%Y/%m/%d %T"
        } else {
            "%d/%m/%Y %T"
        };

        match self {
            Self::Day(date) => date.format(format_date).to_string(),
            Self::DayTime(date) => date.format(format_datetime).to_string(),
            Self::NoDate => String::new(),
        }
    }
}

#[derive(Debug, Hash, Eq, PartialEq, Clone)]
pub struct Task {
    pub subtasks: Vec<Task>,
    pub description: Option<String>,
    pub due_date: DueDate,
    pub line_number: usize,
    pub name: String,
    pub priority: usize,
    pub state: State,
    pub tags: Option<Vec<String>>,
}

impl Default for Task {
    fn default() -> Self {
        Self {
            due_date: DueDate::NoDate,
            name: String::new(),
            priority: 0,
            state: State::ToDo,
            tags: None,
            description: None,
            line_number: 1,
            subtasks: vec![],
        }
    }
}

impl fmt::Display for Task {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        writeln!(f, "Name: \t -> {0}", self.name)?;

        match self.due_date {
            DueDate::NoDate => (),
            DueDate::DayTime(date) => writeln!(f, "Due: \t -> {date}")?,
            DueDate::Day(date) => writeln!(f, "Due: \t {date}")?,
        }

        writeln!(f, "State: \t -> {}", self.state)?;

        if self.priority > 0 {
            writeln!(f, "Prio.: \t -> {}", self.priority)?;
        }

        if let Some(description) = &self.description {
            writeln!(f, "Desc.:\t -> \"{description}\"")?;
        }
        if let Some(tags) = &self.tags {
            writeln!(f, "Tags: \t -> {tags:?}")?;
        }
        fmt::Result::Ok(())
    }
}
impl Task {
    fn get_fixed_attributes(&self, config: &Config, indent_length: usize) -> String {
        let indent = " ".repeat(indent_length);

        let state_str = match self.state {
            State::Done => "X",
            State::ToDo => " ",
        };

        let priority = if self.priority > 0 {
            format!("p{} ", self.priority)
        } else {
            String::new()
        };

        let mut due_date = self
            .due_date
            .to_string_format(!config.tasks_config.use_american_format);
        if !due_date.is_empty() {
            due_date.push(' ');
        }

        let tags_str = self.tags.as_ref().map_or_else(String::new, |tags| {
            tags.clone()
                .iter()
                .map(|t| format!("#{t}"))
                .collect::<Vec<String>>()
                .join(" ")
        });

        let res = format!(
            "{}- [{}] {} {}{}{}",
            indent, state_str, self.name, due_date, priority, tags_str
        );
        res.trim_end().to_string()
    }

    pub fn fix_task_attributes(&self, config: &Config, path: &PathBuf) -> Result<()> {
        let content = read_to_string(path.clone())?;
        let mut lines = content.split('\n').collect::<Vec<&str>>();

        if lines.len() < self.line_number - 1 {
            bail!(
                "Task's line number {} was greater than length of file {:?}",
                self.line_number,
                path
            );
        }

        let indent_length = lines[self.line_number - 1]
            .chars()
            .take_while(|c| c.is_whitespace())
            .count();

        let fixed_line = self.get_fixed_attributes(config, indent_length);

        if lines[self.line_number - 1] != fixed_line {
            debug!(
                "\nReplacing\n{}\nWith\n{}\n",
                lines[self.line_number - 1],
                self.get_fixed_attributes(config, indent_length,)
            );
            lines[self.line_number - 1] = &fixed_line;

            let mut file = File::create(path)?;
            file.write_all(lines.join("\n").as_bytes())?;

            info!("Wrote to {path:?} at line {}", self.line_number);
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use chrono::NaiveDate;

    use crate::{
        config::Config,
        task_core::task::{DueDate, State, Task},
    };

    #[test]
    fn test_fix_attributes() {
        let mut config = Config::default();
        config.tasks_config.use_american_format = true;
        let task = Task {
            due_date: DueDate::Day(NaiveDate::from_ymd_opt(2021, 12, 3).unwrap()),
            name: String::from("Test Task"),
            priority: 1,
            state: State::ToDo,
            tags: Some(vec![String::from("tag1"), String::from("tag2")]),
            description: Some(String::from("This is a test task.")),
            line_number: 2,
            subtasks: vec![],
        };
        let res = task.get_fixed_attributes(&config, 0);
        assert_eq!(res, "- [ ] Test Task 2021/12/03 p1 #tag1 #tag2");
    }

    #[test]
    fn test_fix_attributes_with_no_date() {
        let config = Config::default();
        let task = Task {
            due_date: DueDate::NoDate,
            name: String::from("Test Task with No Date"),
            priority: 2,
            state: State::Done,
            tags: Some(vec![String::from("tag3")]),
            description: None,
            line_number: 3,
            subtasks: vec![],
        };

        let res = task.get_fixed_attributes(&config, 0);
        assert_eq!(res, "- [X] Test Task with No Date p2 #tag3");
    }
}
