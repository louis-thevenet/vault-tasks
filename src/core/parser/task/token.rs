use chrono::{NaiveDate, NaiveTime};

use crate::core::task::State;

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum Token {
    DueDate(NaiveDate),
    DueTime(NaiveTime),
    Name(String),
    Priority(usize),
    Tag(String),
    State(State),
    TodayFlag,
}
