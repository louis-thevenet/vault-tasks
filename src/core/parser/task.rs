mod parse_completion;
mod parse_today;
mod parser_priorities;
mod parser_state;
mod parser_tags;
mod token;

use chrono::NaiveDateTime;
use parse_completion::parse_completion;
use parse_today::parse_today;
use parser_priorities::parse_priority;
use parser_state::parse_task_state;
use parser_tags::parse_tag;
use token::Token;
use tracing::error;
use winnow::{
    Parser, Result,
    combinator::{alt, fail, repeat},
    token::any,
};

use crate::core::{TasksConfig, date::Date, task::Task};

use super::{
    parser_date::{self},
    parser_time,
};

/// Parses a `Token` from an input string.FileEntry
fn parse_token(input: &mut &str, config: &TasksConfig) -> Result<Token> {
    alt((
        // |input: &mut &str| (parse_naive_date(input, config.use_american_format)),
        |input: &mut &str| {
            let date = parser_date::parse_naive_date(input, config.use_american_format)?;
            Ok(Token::DueDate(date))
        },
        |input: &mut &str| {
            let date = parser_time::parse_naive_time(input)?;
            Ok(Token::DueTime(date))
        },
        parse_tag,
        |input: &mut &str| parse_task_state(input, &config.task_state_markers),
        parse_priority,
        parse_completion,
        parse_today,
        |input: &mut &str| {
            let res = repeat(0.., any)
                .fold(String::new, |mut string, c| {
                    string.push(c);
                    string
                })
                .parse_next(input)?;
            Ok(Token::Name(res))
        },
    ))
    .parse_next(input)
}

/// Parses a `Task` from an input string. Filename must be specified to be added to the task.
///
/// # Errors
///
/// Will return an error if the task can't be parsed.
#[allow(clippy::module_name_repetitions)]
pub fn parse_task(input: &mut &str, filename: String, config: &TasksConfig) -> Result<Task> {
    let task_state = match parse_task_state(input, &config.task_state_markers)? {
        Token::State(state) => Ok(state),
        _ => fail(input),
    }?;

    let mut token_parser = |input: &mut &str| parse_token(input, config);

    let tokens = input
        .split_ascii_whitespace()
        .map(|token| token_parser.parse(token));

    let mut task = Task {
        state: task_state,
        filename,
        ..Default::default()
    };

    // Placeholders for a date and a time
    let mut due_date_opt = None;
    let mut due_time_opt = None;
    let mut name_vec = vec![]; // collects words that aren't tokens from the input string

    for token_res in tokens {
        match token_res {
            Ok(Token::DueDate(date)) => due_date_opt = Some(date),
            Ok(Token::DueTime(time)) => due_time_opt = Some(time),
            Ok(Token::Name(name)) => name_vec.push(name),
            Ok(Token::Priority(p)) => task.priority = p,
            Ok(Token::State(state)) => task.state = state,
            Ok(Token::Tag(tag)) => {
                if let Some(ref mut tags) = task.tags {
                    tags.push(tag);
                } else {
                    task.tags = Some(vec![tag]);
                }
            }
            Ok(Token::TodayFlag) => task.is_today = true,
            Ok(Token::CompletionPercentage(c)) => task.completion = Some(c),
            Err(error) => error!("Error: {error:?}"),
        }
    }

    if !name_vec.is_empty() {
        task.name = name_vec.join(" ");
    }
    let due_date = if let Some(due_date) = due_date_opt {
        match due_time_opt {
            Some(time) => Some(Date::DayTime(NaiveDateTime::new(due_date, time))),
            None => Some(Date::Day(due_date)),
        }
    } else {
        None
    };

    task.due_date = due_date;
    Ok(task)
}
#[cfg(test)]
mod test {

    use chrono::{Datelike, Days, NaiveDate, NaiveDateTime, NaiveTime};

    use crate::core::{
        TasksConfig,
        date::Date,
        parser::task::parse_task,
        task::{State, Task},
    };
    #[test]
    fn test_parse_task_no_description() {
        let mut input = "- [x] 10/15 task_name #done";
        let config = TasksConfig {
            use_american_format: true,
            ..Default::default()
        };
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let year = chrono::Local::now().year();
        let expected = Task {
            name: "task_name".to_string(),
            description: None,
            tags: Some(vec!["done".to_string()]),
            due_date: Some(Date::Day(NaiveDate::from_ymd_opt(year, 10, 15).unwrap())),
            priority: 0,
            state: State::Done,
            line_number: Some(1),
            ..Default::default()
        };
        assert_eq!(res, expected);
    }

    #[test]
    fn test_parse_task_only_state() {
        let mut input = "- [ ]";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let expected = Task {
            subtasks: vec![],
            name: String::new(),
            description: None,
            tags: None,
            due_date: None,
            priority: 0,
            state: State::ToDo,
            line_number: Some(1),
            filename: String::new(),
            is_today: false,
            completion: None,
        };
        assert_eq!(res, expected);
    }
    #[test]
    fn test_parse_task_with_due_date_words() {
        let mut input = "- [ ] today 15:30 task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let expected_date = chrono::Local::now().date_naive();
        let expected_time = NaiveTime::from_hms_opt(15, 30, 0).unwrap();
        let expected_due_date = Date::DayTime(NaiveDateTime::new(expected_date, expected_time));
        assert_eq!(res.due_date, Some(expected_due_date));
    }

    #[test]
    fn test_parse_task_with_weekday() {
        let mut input = "- [ ] monday 15:30 task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();

        let now = chrono::Local::now();
        let expected_date = now
            .date_naive()
            .checked_add_days(Days::new(
                8 - u64::from(now.date_naive().weekday().number_from_monday()),
            ))
            .unwrap();
        let expected_time = NaiveTime::from_hms_opt(15, 30, 0).unwrap();
        let expected_due_date = Date::DayTime(NaiveDateTime::new(expected_date, expected_time));
        assert_eq!(res.due_date, Some(expected_due_date));
    }

    #[test]
    fn test_parse_task_with_weekday_this() {
        let mut input = "- [ ] this monday 15:30 task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let now = chrono::Local::now();
        let expected_date = now
            .date_naive()
            .checked_add_days(Days::new(
                8 - u64::from(now.date_naive().weekday().number_from_monday()),
            ))
            .unwrap();
        let expected_time = NaiveTime::from_hms_opt(15, 30, 0).unwrap();
        let expected_due_date = Date::DayTime(NaiveDateTime::new(expected_date, expected_time));
        assert_eq!(res.due_date, Some(expected_due_date));
    }

    #[test]
    fn test_parse_task_with_weekday_next() {
        let mut input = "- [ ] next monday 15:30 task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let now = chrono::Local::now();
        let expected_date = now
            .date_naive()
            .checked_add_days(Days::new(
                8 - u64::from(now.date_naive().weekday().number_from_monday()),
            ))
            .unwrap();
        let expected_time = NaiveTime::from_hms_opt(15, 30, 0).unwrap();
        let expected_due_date = Date::DayTime(NaiveDateTime::new(expected_date, expected_time));
        assert_eq!(res.due_date, Some(expected_due_date));
    }

    #[test]
    fn test_parse_task_without_due_date() {
        let mut input = "- [ ] task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let expected_due_date = None;
        assert_eq!(res.due_date, expected_due_date);
    }

    #[test]
    fn test_parse_task_with_invalid_state() {
        let mut input = "- [invalid] task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_err());
    }

    #[test]
    fn test_parse_task_without_state() {
        let mut input = "task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_err());
    }

    #[test]
    fn test_parse_task_with_invalid_priority() {
        let mut input = "- [ ] task_name p-9";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        assert_eq!(res.priority, 0);
    }

    #[test]
    fn test_parse_task_without_name() {
        let mut input = "- [ ]";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        assert_eq!(res.name, ""); // Default name is used when no name is provided
    }
    #[test]
    fn test_parse_task_with_today_flag() {
        let mut input = "- [ ] @t";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        assert!(res.is_today);
    }
}
