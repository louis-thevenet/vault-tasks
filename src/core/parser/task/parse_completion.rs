use winnow::{combinator::preceded, token::take_while, Parser, Result};

use super::token::Token;

/// Parses a completion percentage of the form `"<integer>%"`.
pub fn parse_completion(input: &mut &str) -> Result<Token> {
    let res = preceded('c', take_while(1.., '0'..='9'))
        .parse_to()
        .parse_next(input)?;

    Ok(Token::CompletionPercentage(res))
}

#[cfg(test)]
mod tests {
    use crate::core::parser::task::{parse_completion::parse_completion, token::Token};

    #[test]
    fn test_parse_completion_success() {
        let mut with_tag = "c99";
        assert_eq!(
            parse_completion(&mut with_tag),
            Ok(Token::CompletionPercentage(99))
        );
    }
    #[test]
    fn test_parse_completion_fail() {
        let mut without_tag = "test";
        assert!(parse_completion(&mut without_tag).is_err());
    }
}
