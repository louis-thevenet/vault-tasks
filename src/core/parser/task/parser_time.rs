use chrono::NaiveTime;
use winnow::{Result, combinator::separated, prelude::*, token::take_while};

use super::token::Token;

/// Parses a `NaiveTime` from a `hh:mm:ss` or `hh:mm` string.
pub fn parse_naive_time(input: &mut &str) -> Result<Token> {
    separated(2..=3, take_while(1.., '0'..='9').parse_to::<u32>(), ':')
        .verify_map(|tokens: Vec<u32>| {
            let h = tokens[0];
            let m = tokens[1];
            let s = if tokens.len() == 3 { tokens[2] } else { 0 };
            if h < 24 && m < 60 && s < 60 {
                Some(Token::DueTime(NaiveTime::from_hms_opt(h, m, s).unwrap()))
            } else {
                None
            }
        })
        .parse_next(input)
}

#[cfg(test)]
mod tests {
    use chrono::{NaiveTime, Timelike};

    use crate::core::parser::task::{parser_time::parse_naive_time, token::Token};

    #[test]
    fn test_parse_wrong_naive_time() {
        let input = "24:50".to_string();
        assert!(parse_naive_time(&mut input.as_str()).is_err());
        let input = "22:90".to_string();
        assert!(parse_naive_time(&mut input.as_str()).is_err());
        let input = "23:50:61".to_string();
        assert!(parse_naive_time(&mut input.as_str()).is_err());
    }
    #[test]
    fn test_parse_naive_time() {
        let now = chrono::Local::now().time();
        let (h, m, s) = (now.hour(), now.minute(), now.second());

        let input = format!("{h}:{m}:{s}");
        let expected = NaiveTime::from_hms_opt(h, m, s).unwrap();
        assert_eq!(
            parse_naive_time(&mut input.as_str()),
            Ok(Token::DueTime(expected))
        );

        let input = format!("{h}:{m}");
        let expected = NaiveTime::from_hms_opt(h, m, 0).unwrap();
        assert_eq!(
            parse_naive_time(&mut input.as_str()),
            Ok(Token::DueTime(expected))
        );
    }
}
