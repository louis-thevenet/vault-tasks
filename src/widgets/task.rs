use ratatui::{
    prelude::*,
    widgets::{Block, Borders, Paragraph},
};

use crate::task_core::task::{State, Task};

pub struct TaskWidget {
    task: Task,
    not_american_format: bool,
}

impl TaskWidget {
    // pub fn new(config: Config, task: Task) -> Self {
    //     Self {
    //         task,
    //         not_american_format: !config.tasks_config.use_american_format,
    //     }
    // }
}
impl Widget for TaskWidget {
    fn render(self, area: ratatui::prelude::Rect, buf: &mut ratatui::prelude::Buffer)
    where
        Self: Sized,
    {
        let state = if self.task.state == State::ToDo {
            "‚ùå"
        } else {
            "‚úÖ"
        };
        let title = Span::styled(format!("{state} {}", self.task.name), Style::default());
        let surrounding_block = Block::default().borders(Borders::ALL).title(title.clone());

        let mut lines = vec![];

        let mut data_line = String::new();
        let due_date_str = self
            .task
            .due_date
            .to_string_format(self.not_american_format);

        if !due_date_str.is_empty() {
            data_line.push_str(&format!("üìÖ {due_date_str} "));
        }
        if self.task.priority > 0 {
            data_line.push_str(&format!("‚ùó{} ", self.task.priority));
        }
        if !data_line.is_empty() {
            lines.push(Line::from(Span::styled(data_line, Style::default())));
        }
        let mut tag_line = String::new();
        if self.task.tags.is_some() {
            tag_line.push_str(
                &self
                    .task
                    .tags
                    .unwrap()
                    .iter()
                    .map(|t| format!("#{t}"))
                    .collect::<Vec<String>>()
                    .join(" "),
            );
        }
        if !tag_line.is_empty() {
            lines.push(Line::from(Span::styled(tag_line, Color::DarkGray)));
        }
        let description = self.task.description.unwrap_or_default();
        if !description.is_empty() {
            description
                .split('\n')
                .map(|l| Line::from(Span::styled(l, Color::Gray)))
                .for_each(|l| {
                    lines.push(l);
                });
        }

        if lines.is_empty() {
            let p = Paragraph::new(title);
            p.render(area, buf);
        } else {
            let text = Text::from(lines);

            let p = Paragraph::new(text).block(surrounding_block);
            p.render(area, buf);
        }
    }
}
