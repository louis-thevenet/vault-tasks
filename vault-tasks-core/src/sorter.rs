use std::cmp::Ordering;

use chrono::NaiveTime;
use lexical_sort::lexical_cmp;
use strum::EnumIter;
use strum_macros::FromRepr;

use super::task::{DueDate, Task};

#[derive(Default, Clone, Copy, FromRepr, EnumIter, strum_macros::Display)]
pub enum SortingMode {
    #[default]
    #[strum(to_string = "Due Date")]
    ByDueDate,
    #[strum(to_string = "Title")]
    ByName,
}

impl SortingMode {
    pub fn next(self) -> Self {
        match self {
            SortingMode::ByDueDate => SortingMode::ByName,
            SortingMode::ByName => SortingMode::ByDueDate,
        }
    }
    pub fn sort(tasks: &mut [Task], sorter: SortingMode) {
        tasks.sort_by(|t1, t2| Self::cmp(t1, t2, sorter));
    }

    /// Compare two tasks by due date
    fn cmp_due_date(t1: &Task, t2: &Task) -> Ordering {
        match (&t1.due_date, &t2.due_date) {
            (DueDate::Day(d1), DueDate::Day(d2)) => d1.cmp(d2),
            (DueDate::DayTime(d1), DueDate::DayTime(d2)) => d1.cmp(d2),
            (DueDate::Day(d1), DueDate::DayTime(d2)) => d1.and_time(NaiveTime::default()).cmp(d2),
            (DueDate::DayTime(d1), DueDate::Day(d2)) => d1.cmp(&d2.and_time(NaiveTime::default())),
            (DueDate::NoDate, DueDate::Day(_) | DueDate::DayTime(_)) => Ordering::Greater,
            (DueDate::Day(_) | DueDate::DayTime(_), DueDate::NoDate) => Ordering::Less,
            _ => Ordering::Equal,
        }
    }
    /// Compares two tasks with the specified sorting mode
    /// Sorting mode is used first
    /// If equal, other attribues will be used:
    /// - State: `ToDo` < `Done` (in Ord impl of `State`)
    /// - The other sorting mode
    /// - Priority: usual number ordering
    /// - Tags: not used
    fn cmp(t1: &Task, t2: &Task, sorter: SortingMode) -> Ordering {
        let res_initial_sort = match sorter {
            SortingMode::ByDueDate => Self::cmp_due_date(t1, t2),
            SortingMode::ByName => lexical_cmp(&t1.name, &t2.name),
        };

        if !matches!(res_initial_sort, Ordering::Equal) {
            return res_initial_sort;
        }

        // Compare states
        let res = t1.state.cmp(&t2.state);
        if !matches!(res, Ordering::Equal) {
            return res;
        }

        // We do the other sorting methods
        let res = match sorter {
            SortingMode::ByDueDate => lexical_cmp(&t1.name, &t2.name),
            SortingMode::ByName => Self::cmp_due_date(t1, t2),
        };
        if !matches!(res, Ordering::Equal) {
            return res;
        }

        t1.priority.cmp(&t2.priority)
    }
}
#[cfg(test)]
mod tests {

    use insta::{assert_debug_snapshot, with_settings};

    use super::SortingMode;
    use crate::{parser::task::parse_task, task::Task, TasksConfig};
    #[test]
    fn task_sort_by_name() {
        let mut source = [
            "- [ ] test 10/11",
            "- [ ] test 10/9",
            "- [ ] test 10/10 p5",
            "- [ ] test 10/10 10:00",
            "- [x] zèbre",
            "- [x] zzz",
            "- [ ] zzz",
            "- [ ] test 10/10 p2",
            "- [x] test",
            "- [ ] test2",
            "- [ ] test 10/10 5:00",
            "- [ ] abc",
        ];
        let config = TasksConfig {
            use_american_format: true,
            task_state_marker: 'x',
            ..Default::default()
        };
        let mut tasks: Vec<Task> = source
            .iter_mut()
            .map(|input| parse_task(input, String::new(), &config).unwrap())
            .collect();

        let sorting_mode = SortingMode::ByName;
        SortingMode::sort(&mut tasks, sorting_mode);

        let tasks = tasks
            .iter()
            .map(|task| task.get_fixed_attributes(&config, 2))
            .collect::<Vec<String>>();

        with_settings!({
            info=>&source,
            description => "", // the template source code
        }, {
                assert_debug_snapshot!(tasks);
        });
    }
    #[test]
    fn task_sort_by_due_date() {
        let mut source = [
            "- [ ] test 10/11",
            "- [ ] test 10/9",
            "- [ ] test 10/10 p5",
            "- [ ] test 10/10 10:00",
            "- [x] zèbre",
            "- [x] zzz",
            "- [ ] zzz",
            "- [ ] test 10/10 p2",
            "- [x] test",
            "- [ ] test2",
            "- [ ] test 10/10 5:00",
            "- [ ] abc",
        ];
        let config = TasksConfig {
            use_american_format: true,
            task_state_marker: 'x',
            ..Default::default()
        };
        let mut tasks: Vec<Task> = source
            .iter_mut()
            .map(|input| parse_task(input, String::new(), &config).unwrap())
            .collect();

        let sorting_mode = SortingMode::ByDueDate;
        SortingMode::sort(&mut tasks, sorting_mode);

        let tasks = tasks
            .iter()
            .map(|task| task.get_fixed_attributes(&config, 2))
            .collect::<Vec<String>>();

        with_settings!({
            info=>&source,
            description => "", // the template source code
        }, {
                assert_debug_snapshot!(tasks);
        });
    }
}
