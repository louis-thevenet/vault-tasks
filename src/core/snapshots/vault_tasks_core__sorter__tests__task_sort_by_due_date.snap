---
source: vault-tasks-core/src/sorter.rs
description: ""
expression: tasks
info:
  - " test 10/11"
  - " test 10/9"
  - " test 10/10 p5"
  - " test 10/10 10:00"
  - " zèbre"
  - " zzz"
  - " zzz"
  - " test 10/10 p2"
  - " test"
  - " test2"
  - " test 10/10 5:00"
  - " abc"
---
[
    "  - [ ] test 2024/10/09",
    "  - [ ] test 2024/10/10 p2",
    "  - [ ] test 2024/10/10 p5",
    "  - [ ] test 2024/10/10 05:00:00",
    "  - [ ] test 2024/10/10 10:00:00",
    "  - [ ] test 2024/10/11",
    "  - [ ] abc",
    "  - [ ] test2",
    "  - [ ] zzz",
    "  - [x] test",
    "  - [x] zèbre",
    "  - [x] zzz",
]
